src/machine/machine_k210.go

```diff
@@ -0,0 +1,642 @@
+// +build k210
+
+package machine
+
+import (
+	"device/kendryte"
+	"errors"
+	"runtime/interrupt"
+)
+
+func CPUFrequency() uint32 {
+	return 390000000
+}
+
+type PinMode uint8
+type fpioaPullMode uint8
+type PinChange uint8
+
+// Pin modes.
+const (
+	PinInput PinMode = iota
+	PinInputPullUp
+	PinInputPullDown
+	PinOutput
+)
+
+// FPIOA internal pull resistors.
+const (
+	fpioaPullNone fpioaPullMode = iota
+	fpioaPullDown
+	fpioaPullUp
+)
+
+// GPIOHS pin interrupt events.
+const (
+	PinRising PinChange = iota + 1
+	PinFalling
+	PinToggle
+	PinHigh
+	PinLow = 8
+)
+
+var (
+	errUnsupportedSPIController = errors.New("SPI controller not supported. Use SPI0 or SPI1.")
+	errI2CTxAbort               = errors.New("I2C transmition has been aborted.")
+)
+
+func (p Pin) setFPIOAIOPull(pull fpioaPullMode) {
+	switch pull {
+	case fpioaPullNone:
+		kendryte.FPIOA.IO[uint8(p)].ClearBits(kendryte.FPIOA_IO_PU & kendryte.FPIOA_IO_PD)
+	case fpioaPullUp:
+		kendryte.FPIOA.IO[uint8(p)].SetBits(kendryte.FPIOA_IO_PU)
+		kendryte.FPIOA.IO[uint8(p)].ClearBits(kendryte.FPIOA_IO_PD)
+	case fpioaPullDown:
+		kendryte.FPIOA.IO[uint8(p)].ClearBits(kendryte.FPIOA_IO_PU)
+		kendryte.FPIOA.IO[uint8(p)].SetBits(kendryte.FPIOA_IO_PD)
+	}
+}
+
+// SetFPIOAFunction is used to configure the pin for one of the FPIOA functions.
+// Each pin on the Kendryte K210 can be configured with any of the available FPIOA functions.
+func (p Pin) SetFPIOAFunction(f FPIOAFunction) {
+	kendryte.FPIOA.IO[uint8(p)].Set(fpioaFuncDefaults[uint8(f)])
+}
+
+// FPIOAFunction returns the current FPIOA function of the pin.
+func (p Pin) FPIOAFunction() FPIOAFunction {
+	return FPIOAFunction((kendryte.FPIOA.IO[uint8(p)].Get() & kendryte.FPIOA_IO_CH_SEL_Msk))
+}
+
+// Configure this pin with the given configuration.
+// The pin must already be set as GPIO or GPIOHS pin.
+func (p Pin) Configure(config PinConfig) {
+	var input bool
+
+	// Check if the current pin's FPIOA function is either GPIO or GPIOHS.
+	f := p.FPIOAFunction()
+	if f < FUNC_GPIOHS0 || f > FUNC_GPIO7 {
+		return // The pin is not configured as GPIO or GPIOHS.
```

Comment:
I would expect `Configure` to set up the pin as a GPIO pin, even if it is already configured as something else (so calling `Configure` is a way to undo a different configuration). Would it be possible to reconfigure this pin as a GPIO pin, or is there a reason not to do that?

If the pin is again used by a different peripheral (such as SPI), that peripheral should set the correct pin mode for use by that peripheral.

Reply:
@aykevl Yes it would be possible to configure a non-GPIO pin as GPIO automatically. However I think it is better to assign a none default functions to a pin using the `SetFPIOAFunction()`. Imagine that situation:
```go
machine.D0.Configure(machine.PinConfig{Mode: PinOutput}) // D0 has a JTAG function as default, imagine we configure it automatically as GPIOHS0
machine.D1.Configure(machine.PinConfig{Mode: PinInputPullNone}) // D1 has also a JTAG function as default, but we also configure it as GPIOHS0
machine.D0.High() // This would have no effect now because D1 configured GPIOHS0 as input pin and D0 also has the GPIOHS0 function.
```
So this would lead to confusing situations. I know the `SetFPIOAFunction()` adds a step before you can use a certain pin (unless you use a pin with a default function which is already GPIO, see `board_maixbit.go`) but at least there is no risk that two pins have the same function unless you do it on purpose. What do you think?

Reply:
Does that mean that with the PR as-is, you also need to call `SetFPIOAFunction` to do anything with these pins? Which implies that examples/blinky1 for example won't work?

Reply:
Also, am I correct that you can actually use the debug pins for something else? That seems dangerous, I'm not sure that should be allowed.

Reply:
No, if the default FPIOA function for the pin is already GPIO or GPIOHS you do not have to call `SetFPIOAFunction` (the led pins `D12`,`D13`,`D14` are GPIO as default). This would only be needed for the pins `D0` - `D7` or the K210 pins `P36`-`P47` (although they aren't even mapped to external pins on the MAix Bit).

Reply:
> Also, am I correct that you can actually use the debug pins for something else? That seems dangerous, I'm not sure that should be allowed.

Yes you can, but you would have to do a `SetFPIOAFunction`.

Reply:
Ok, I think we can leave it as-is for now. Cross-platform code can simply consider these pins to have a fixed function (like AVR chips for example). I'm not so sure about exposing `SetFPIOAFunction` and `FPIOAFunction`, they add chip specific changes that technically are not necessary (`Configure` calls can do the same, and in fact do just that on most other chips such as SAM D chips).